### PR TITLE
Derivative metadata

### PR DESCRIPTION
**Pull request by sarmatdev**
- Add `tokenURI` function to the `Derivative` base contract
- Add tests

### CODE REVIEW

File: contracts/base/Derivative.sol [hunk 1]
```diff
@@ -107,6 +110,39 @@ contract Derivative is ERC721, Ownable, Versioned {
...
+  }
+
+  function _baseURI() internal view virtual override returns (string memory) {
+    return
+      bytes(derivativeBaseURI).length > 0
+        ? derivativeBaseURI
+        : ILedger(ledgerContract).getBaseURI();
+  }
+
+  function setBaseURI(string memory _newBaseURI) external onlyOwner {
```
Comment (backmeupplz) [inline contracts/base/Derivative.sol:142]: There is no need in prefixing function argument names with `new`. `_baseURI` is enough to distinguish it from the `baseURI` instance variable for the contract

File: contracts/base/Derivative.sol [hunk 2]
```diff
@@ -62,17 +62,20 @@ pragma solidity ^0.8.12;
...
 import "@openzeppelin/contracts/access/Ownable.sol";
 import "@openzeppelin/contracts/utils/Counters.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
 import "@big-whale-labs/versioned-contract/contracts/Versioned.sol";
+import "../interfaces/ILedger.sol";
 
 contract Derivative is ERC721, Ownable, Versioned {
   using Counters for Counters.Counter;
-
+  using Strings for uint256;
```
Comment (backmeupplz) [inline contracts/base/Derivative.sol:71]: Did we lose the empty line between `using`'s and `// State` here?

**Review by backmeupplz — CHANGES_REQUESTED** [review]

File: contracts/base/Derivative.sol [hunk 3]
```diff
@@ -107,6 +110,39 @@ contract Derivative is ERC721, Ownable, Versioned {
...
+    public
+    view
+    virtual
+    override
+    returns (string memory)
+  {
+    _requireMinted(tokenId);
+
+    string memory baseURI = _baseURI();
+    string memory contractAddress = Strings.toHexString(
```
[the remark this answers is not in the document]
Comment (sarmatdev) [reply contracts/base/Derivative.sol:123]: `abi.encodePacked` returns bytes, but even if we cast the to string, we don't get the address itself

File: contracts/base/Derivative.sol [hunk 4]
```diff
@@ -107,6 +112,31 @@ contract Derivative is ERC721, Ownable, Versioned {
...
 
+  function tokenURI(uint256 tokenId)
+    public
+    view
+    virtual
+    override
+    returns (string memory)
+  {
+    _requireMinted(tokenId);
+    require(bytes(baseURI).length > 0, "baseURI does not set");
```
Comment (backmeupplz) [inline contracts/base/Derivative.sol:123]: `baseURI isn't set`

**Review by backmeupplz — APPROVED** [review]

**Pull request merged**